routers/web/repo/treelist.go

```diff
@@ -0,0 +1,55 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package repo
+
+import (
+	"net/http"
+
+	"code.gitea.io/gitea/modules/base"
+	"code.gitea.io/gitea/modules/context"
+	"code.gitea.io/gitea/modules/git"
+
+	"github.com/go-enry/go-enry/v2"
+)
+
+// TreeList get all files' entries of a repository
+func TreeList(ctx *context.Context) {
+	tree, err := ctx.Repo.Commit.SubTree("/")
+	if err != nil {
+		ctx.ServerError("Repo.Commit.SubTree", err)
+		return
+	}
+
+	entries, err := tree.ListEntriesRecursive()
```

Comment:
That's what GitHub does at the moment. And see my comment.

![image](https://user-images.githubusercontent.com/2114189/172408899-fe4cbcb5-8c53-4ae3-9420-283e905b2df3.png)

Reply:
I think GitHub has its own methods to do the optimization, instead of generating the cache when a commit is pushed to the branch, because the URL `https://github.com/{org}/{repo}/tree-list/{commit-id}` can get the file list for any commit id.

About performance:
* on `gitea` repo (MacBook SSD): `git ls-tree -t -l -r main > /dev/null  0.03s user 0.02s system 93% cpu 0.055 total`
* on `linux` repo (VPS with virtual disk around 100MB/s): `git ls-tree -t -l -r master > /dev/null  0.74s user 0.73s system 29% cpu 5.049 total`, not too bad, it will be much faster on a SSD.

To introduce a cache, I think it's out of the scope of this PR, and in my mind the cache data had better to be put inside the bare git repository on disk (eg: `/data/git/repositories/org/repo.git/gitea-cache/xxxxxx`) because the in-memory doesn't fit the file list data well.